include/cuda/std/detail/libcxx/include/__concepts/swappable.h

```diff
@@ -0,0 +1,225 @@
+    }
+
+    // 2.2   Otherwise, if `E1` and `E2` are lvalues of array types with equal extent and...
+    template<class _Tp, class _Up, size_t _Size>
+      requires __swappable_arrays<_Tp, _Up, _Size>
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr void operator()(_Tp(& __t)[_Size], _Up(& __u)[_Size]) const
+      noexcept(noexcept((*this)(*__t, *__u)))
+    {
+      // TODO(cjdb): replace with `ranges::swap_ranges`.
+      for (size_t __i = 0; __i < _Size; ++__i) {
+        (*this)(__t[__i], __u[__i]);
+      }
+    }
+
+    // 2.3   Otherwise, if `E1` and `E2` are lvalues of the same type `T` that models...
+    template<__exchangeable _Tp>
+    _LIBCUDACXX_INLINE_VISIBILITY constexpr void operator()(_Tp& __x, _Tp& __y) const
+      noexcept(is_nothrow_move_constructible_v<_Tp> && is_nothrow_move_assignable_v<_Tp>)
+    {
+      __y = _CUDA_VSTD::exchange(__x, _CUDA_VSTD::move(__y));
+    }
+  };
+} // namespace __swap
+
+inline namespace __cpo {
+  inline constexpr auto swap = __swap::__fn{};
+} // namespace __cpo
+} // namespace ranges
+
+template<class _Tp>
+concept swappable = requires(_Tp& __a, _Tp& __b) { ranges::swap(__a, __b); };
+
+template<class _Tp, class _Up>
+concept swappable_with =
+  common_reference_with<_Tp, _Up> &&
+  requires(_Tp&& __t, _Up&& __u) {
+    ranges::swap(_CUDA_VSTD::forward<_Tp>(__t), _CUDA_VSTD::forward<_Tp>(__t));
+    ranges::swap(_CUDA_VSTD::forward<_Up>(__u), _CUDA_VSTD::forward<_Up>(__u));
+    ranges::swap(_CUDA_VSTD::forward<_Tp>(__t), _CUDA_VSTD::forward<_Up>(__u));
+    ranges::swap(_CUDA_VSTD::forward<_Up>(__u), _CUDA_VSTD::forward<_Tp>(__t));
+  };
+
+#else if _LIBCUDACXX_STD_VER > 11
+
+namespace ranges {
+namespace __swap {
+
+  template<class _Tp>
+  void swap(_Tp&, _Tp&) = delete;
+  
+  template<class _Tp, class _Up>
+  _LIBCUDACXX_CONCEPT_FRAGMENT(
+    __unqualified_swappable_with_,
+    requires(_Tp&& __t, _Up&& __u) //
+    (swap(_CUDA_VSTD::forward<_Tp>(__t), _CUDA_VSTD::forward<_Up>(__u))));
+
+  template<class _Tp, class _Up>
+  _LIBCUDACXX_CONCEPT __unqualified_swappable_with =
+    (__class_or_enum<remove_cvref_t<_Tp>> || __class_or_enum<remove_cvref_t<_Up>>) &&
+    _LIBCUDACXX_FRAGMENT(__unqualified_swappable_with_, _Tp, _Up);
+
+  struct __fn;
+  
+  template<class _Tp, class _Up, class _Size>
+  _LIBCUDACXX_CONCEPT_FRAGMENT(
+    __swappable_arrays_,
+    requires(_Tp(& __t)[_Size::value], _Up(& __u)[_Size::value], const __fn& __swap) //
+    (__swap(__t[0], __u[0])));
+
+  template<class _Tp, class _Up, size_t _Size>
+  _LIBCUDACXX_CONCEPT __swappable_arrays =
+    !__unqualified_swappable_with<_Tp(&)[_Size], _Up(&)[_Size]> &&
+    extent_v<_Tp> == extent_v<_Up> &&
+    _LIBCUDACXX_FRAGMENT(__swappable_arrays_, _Tp, _Up, _CUDA_VSTD::integral_constant<size_t, _Size>);
+
+  template<class _Tp>
+  _LIBCUDACXX_CONCEPT __exchangeable =
+    !__unqualified_swappable_with<_Tp&, _Tp&> &&
+    move_constructible<_Tp> &&
+    assignable_from<_Tp&, _Tp>;
+
+  struct __fn {
```

Comment:
Yeah, I can do that

Reply:
Actually now that I think about it, I _think_ that ODR will still not be satisfied because we emit different tokens in different language modes. I am not sure if we can do something that isn't technically illegal here like that. Plus the signatures of the functions will differ...

I think an actual way out of the ODR issue is to have the types have different names, i.e. put them in different namespaces between different language modes. This way having both definitions in a program built across 17 and 20 will be fine, and any users doing insane things like passing these across interface boundaries will get a link error (as they should). We should remember this for future conceptified CPOs we add.

Reply:
I think we need to double check, from what I can see the only relevant difference are the constrained member functions and they should not be mangled differently

Reply:
Constraints are a part of a function signature, the fact they aren't mangled in on current implementations is an ABI bug.

Reply:
🙀